File: packages/server/src/auth/manager.ts
import type { FastifyReply, FastifyRequest } from "fastify"
import path from "path"
import type { Logger } from "../logger"
import { AuthStore } from "./auth-store"
import { TokenManager } from "./token-manager"
import { SessionManager } from "./session-manager"
import { isLoopbackAddress, parseCookies } from "./http-auth"

export const BOOTSTRAP_TOKEN_STDOUT_PREFIX = "EMBEDDEDCOWORK_BOOTSTRAP_TOKEN:" as const
export const DEFAULT_AUTH_USERNAME = "embeddedcowork" as const
export const DEFAULT_AUTH_COOKIE_NAME = "embeddedcowork_session" as const

export interface AuthManagerInit {
  configPath: string
  username: string
  password?: string
  generateToken: boolean
  dangerouslySkipAuth?: boolean
  cookieName?: string
}

export class AuthManager {
  private readonly authStore: AuthStore | null
  private readonly tokenManager: TokenManager | null
  private readonly sessionManager = new SessionManager()
  private readonly cookieName: string
  private readonly authEnabled: boolean

  constructor(private readonly init: AuthManagerInit, private readonly logger: Logger) {
    this.cookieName = sanitizeCookieName(init.cookieName)
    this.authEnabled = !Boolean(init.dangerouslySkipAuth)

    if (!this.authEnabled) {
      this.authStore = null
      this.tokenManager = null
      return
    }

    const authFilePath = resolveAuthFilePath(init.configPath)
    this.authStore = new AuthStore(authFilePath, logger.child({ component: "auth" }))

    // Startup: password comes from CLI/env, auth.json, or bootstrap-only mode.
    this.authStore.ensureInitialized({
      username: init.username,
      password: init.password,
      allowBootstrapWithoutPassword: init.generateToken,
    })

    this.tokenManager = init.generateToken ? new TokenManager(60_000) : null
  }

  isAuthEnabled(): boolean {
    return this.authEnabled
  }

  getCookieName(): string {
    return this.cookieName
  }

  isTokenBootstrapEnabled(): boolean {
    return Boolean(this.tokenManager)
  }

  issueBootstrapToken(): string | null {
    if (!this.tokenManager) return null
    return this.tokenManager.generate()
  }

  consumeBootstrapToken(token: string): boolean {
    if (!this.tokenManager) return false
    return this.tokenManager.consume(token)
  }

  validateLogin(username: string, password: string): boolean {
    if (!this.authEnabled) {
      return true
    }
    return this.requireAuthStore().validateCredentials(username, password)
  }

  createSession(username: string) {
    if (!this.authEnabled) {
      return { id: "auth-disabled", createdAt: Date.now(), username: this.init.username }
    }
    return this.sessionManager.createSession(username)
  }

  getStatus() {
    if (!this.authEnabled) {
      return { username: this.init.username, passwordUserProvided: false }
    }
    return this.requireAuthStore().getStatus()
  }

  setPassword(password: string) {
    if (!this.authEnabled) {
      throw new Error("Internal authentication is disabled")
    }
    return this.requireAuthStore().setPassword({ password, markUserProvided: true })
  }

  isLoopbackRequest(request: FastifyRequest): boolean {
    return isLoopbackAddress(request.socket.remoteAddress)
  }

  getSessionFromRequest(request: FastifyRequest): { username: string; sessionId: string } | null {
    return this.getSessionFromHeaders(request.headers)
  }

  getSessionFromHeaders(headers: { cookie?: string | string[] | undefined }): { username: string; sessionId: string } | null {
    if (!this.authEnabled) {
      // When auth is disabled, treat all requests as authenticated.
      // We still return a stable username so callers can display it.
      return { username: this.init.username, sessionId: "auth-disabled" }
    }

    const cookieHeader = Array.isArray(headers.cookie) ? headers.cookie.join("; ") : headers.cookie
    const cookies = parseCookies(cookieHeader)
    const sessionId = cookies[this.cookieName]
    const session = this.sessionManager.getSession(sessionId)
    if (!session) return null
    return { username: session.username, sessionId: session.id }
  }

  setSessionCookie(reply: FastifyReply, sessionId: string) {
    reply.header("Set-Cookie", buildSessionCookie(this.cookieName, sessionId))
  }

  setSessionCookieWithOptions(reply: FastifyReply, sessionId: string, options?: { secure?: boolean }) {
    reply.header("Set-Cookie", buildSessionCookie(this.cookieName, sessionId, options))
  }

  clearSessionCookie(reply: FastifyReply) {
    reply.header("Set-Cookie", buildSessionCookie(this.cookieName, "", { maxAgeSeconds: 0 }))
  }

  clearSessionCookieWithOptions(reply: FastifyReply, options?: { secure?: boolean }) {
    reply.header("Set-Cookie", buildSessionCookie(this.cookieName, "", { maxAgeSeconds: 0, ...options }))
  }

  private requireAuthStore(): AuthStore {
    if (!this.authStore) {
      throw new Error("Auth store is unavailable")
    }
    return this.authStore
  }
}

function sanitizeCookieName(value: string | undefined): string {
  const trimmed = value?.trim()
  if (!trimmed) {
    return DEFAULT_AUTH_COOKIE_NAME
  }

  const sanitized = trimmed.replace(/[^A-Za-z0-9_-]/g, "_")
  return sanitized.length > 0 ? sanitized : DEFAULT_AUTH_COOKIE_NAME
}

function resolveAuthFilePath(configPath: string) {
  const resolvedConfigPath = resolvePath(configPath)
  return path.join(path.dirname(resolvedConfigPath), "auth.json")
}

function resolvePath(filePath: string) {
  if (filePath.startsWith("~/")) {
    return path.join(process.env.HOME ?? "", filePath.slice(2))
  }
  return path.resolve(filePath)
}

function buildSessionCookie(name: string, value: string, options?: { maxAgeSeconds?: number; secure?: boolean }) {
  const parts = [`${name}=${encodeURIComponent(value)}`, "HttpOnly", "Path=/", "SameSite=Lax"]
  if (options?.secure) {
    parts.push("Secure")
  }
  if (options?.maxAgeSeconds !== undefined) {
    parts.push(`Max-Age=${Math.max(0, Math.floor(options.maxAgeSeconds))}`)
  }
  return parts.join("; ")
}

File: packages/server-rust/src/auth/manager.rs

use super::auth_store::{AuthStore, AuthStatus as AuthStoreStatus, SetPasswordParams};
use super::http_auth::{is_loopback_address, parse_cookies};
use super::session_manager::{SessionInfo as SessionManagerSessionInfo, SessionManager};
use super::token_manager::TokenManager;
use crate::logger::Logger;

pub const BOOTSTRAP_TOKEN_STDOUT_PREFIX: &str = "EMBEDDEDCOWORK_BOOTSTRAP_TOKEN:";
pub const DEFAULT_AUTH_USERNAME: &str = "embeddedcowork";
pub const DEFAULT_AUTH_COOKIE_NAME: &str = "embeddedcowork_session";

#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub id: String,
    pub created_at: u64,
    pub username: String,
}

impl From<SessionManagerSessionInfo> for SessionInfo {
    fn from(s: SessionManagerSessionInfo) -> Self {
        SessionInfo {
            id: s.id,
            created_at: s.created_at,
            username: s.username,
        }
    }
}

#[derive(Debug, Clone)]
pub struct AuthStatus {
    pub username: String,
    pub password_user_provided: bool,
}

impl From<AuthStoreStatus> for AuthStatus {
    fn from(s: AuthStoreStatus) -> Self {
        AuthStatus {
            username: s.username,
            password_user_provided: s.password_user_provided,
        }
    }
}

pub struct AuthManager {
    auth_store: Option<AuthStore>,
    token_manager: Option<TokenManager>,
    session_manager: SessionManager,
    cookie_name: String,
    auth_enabled: bool,
    init_username: String,
    #[allow(dead_code)]
    init_password: Option<String>,
    #[allow(dead_code)]
    generate_token: bool,
    pub config_path: String,
    #[allow(dead_code)]
    logger: Logger,
}

impl AuthManager {
    pub fn new(
        auth_store: AuthStore,
        session_manager: SessionManager,
        token_manager: Option<TokenManager>,
    ) -> Self {
        Self {
            auth_store: Some(auth_store),
            session_manager,
            token_manager,
            cookie_name: DEFAULT_AUTH_COOKIE_NAME.to_string(),
            auth_enabled: true,
            init_username: DEFAULT_AUTH_USERNAME.to_string(),
            init_password: None,
            generate_token: false,
            config_path: String::new(),
            logger: Logger::new("auth"),
        }
    }

    pub fn config_path(&self) -> &str {
        &self.config_path
    }

    pub fn apply_init(&mut self, init: AuthManagerInit) {
        self.config_path = init.config_path;
        if !init.username.is_empty() {
            self.init_username = init.username;
        }
        self.init_password = init.password;
        self.generate_token = init.generate_token;
        self.auth_enabled = !init.dangerously_skip_auth;
        if let Some(name) = init.cookie_name {
            if !name.is_empty() {
                self.cookie_name = name;
            }
        }
    }

    pub fn is_auth_enabled(&self) -> bool {
        self.auth_enabled
    }

    pub fn get_cookie_name(&self) -> &str {
        &self.cookie_name
    }

    pub fn is_token_bootstrap_enabled(&self) -> bool {
        self.token_manager.is_some()
    }

    pub fn issue_bootstrap_token(&mut self) -> Option<String> {
        self.token_manager.as_mut().map(|tm| tm.generate())
    }

    pub fn consume_bootstrap_token(&mut self, token: &str) -> bool {
        self.token_manager.as_mut().map_or(false, |tm| tm.consume(token))
    }

    pub fn validate_login(&mut self, username: &str, password: &str) -> bool {
        if !self.auth_enabled {
            return true;
        }
        self.require_auth_store().validate_credentials(username, password)
    }

    pub fn create_session(&mut self, username: &str) -> SessionInfo {
        if !self.auth_enabled {
            return SessionInfo {
                id: "auth-disabled".to_string(),
                created_at: current_time_millis(),
                username: self.init_username.clone(),
            };
        }
        self.session_manager.create_session(username).into()
    }

    pub fn get_status(&mut self) -> AuthStatus {
        if !self.auth_enabled {
            return AuthStatus {
                username: self.init_username.clone(),
                password_user_provided: false,
            };
        }
        let status = self.require_auth_store().get_status();
        status.unwrap_or(AuthStoreStatus {
            username: self.init_username.clone(),
            password_user_provided: false,
        }).into()
    }

    pub fn set_password(&mut self, password: &str) -> Result<AuthStatus, String> {
        if !self.auth_enabled {
            return Err("Internal authentication is disabled".to_string());
        }
        self.require_auth_store().set_password(&SetPasswordParams {
            password: password.to_string(),
            mark_user_provided: true,
        }).map(|s| s.into())
    }

    pub fn is_loopback_request(&self, remote_addr: Option<&str>) -> bool {
        is_loopback_address(remote_addr)
    }

    pub fn get_session_from_headers(&mut self, headers: &HttpHeaders) -> Option<SessionInfo> {
        if !self.auth_enabled {
            return Some(SessionInfo {
                username: self.init_username.clone(),
                id: "auth-disabled".to_string(),
                created_at: current_time_millis(),
            });
        }

        let cookie_header = headers.cookie.as_deref();
        let cookies = parse_cookies(cookie_header);
        let session_id = cookies.get(&self.cookie_name)?;
        let session = self.session_manager.get_session(Some(session_id))?;
        Some(session.clone().into())
    }

    fn require_auth_store(&mut self) -> &mut AuthStore {
        self.auth_store.as_mut().expect("Auth store is unavailable")
    }
}

pub struct AuthManagerInit {
    pub config_path: String,
    pub username: String,
    pub password: Option<String>,
    pub generate_token: bool,
    pub dangerously_skip_auth: bool,
    pub cookie_name: Option<String>,
}

pub struct HttpHeaders {
    pub cookie: Option<String>,
}

fn current_time_millis() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

